### PR TITLE
Improve GLSL language support, navigation, highlighting and formatting

### PR DESCRIPTION
**Pull request by Luna5ama**
## Summary

This PR bundles a number of fixes and improvements accumulated in my fork over the last few years.

The main focus areas are:

* Semantic highlighting
* Macro resolution and navigation
* Include/declaration resolution
* Completion improvements
* Formatter improvements
* GLSL language updates and parser fixes

I cleaned up the commit history and removed experimental/reverted changes before submitting this PR.

---

## Semantic Highlighting

Improved semantic highlighting support for GLSL symbols.

### Added highlighting for

* Function declarations
* Variable declarations
* Field selections
* Additional GLSL identifier categories
* Uniform / in / out / varying / attribute declarations
* Additional GLSL scalar and integer types

This makes symbol highlighting more consistent with modern IDE language support and improves readability in large shader codebases.

---

## Macro Resolution and Navigation

Improved support for macros across project boundaries.

### Added

* Macro reference highlighting
* Global macro reference resolution
* Goto Declaration support for macros

### Fixed

* Resolution of macros defined in included files
* Various edge cases involving include chains

Additional tests were added to cover included macro scenarios.

---

## Include Resolution and Code Navigation

Improved include handling and declaration lookup.

### Fixed

* Include path resolution
* Declaration resolution in several edge cases
* Source root handling
* Uses source root if configured, with fallback to content root

These fixes improve navigation in shader projects that do not follow IntelliJ source-root conventions.

---

### Fixed

* Completion for overloaded functions
* Completion for defines from included files
* Completion for structs and interface blocks from included files

### Cleanup

* Removed an unused encountered-function tracking field from completion logic

---

### Updated

* Built-in gl_PerVertex definitions for additional shader stages

### Improved

* Interface block parsing
* Integer scalar type handling
* Support for newer GLSL type variants （8-bit ints, 16-bit ints, 16-bit floats)

---

### Added

* Function declaration parameter formatting
* Continuation indentation for multiline conditions
* Operator spacing rules
* Unary operator spacing rules

### Fixed

* Condition indentation
* Multiline call indentation
* Switch-case indentation
* Comment spacing preservation

---

## Testing

The PR includes updates and additions to the existing test suite covering:

* Semantic highlighting
* Macro resolution
* Include handling
* Formatter behavior

I have been using these changes in production shader projects for an extended period and have not encountered significant regressions.

---

## Notes

This PR intentionally combines multiple related improvements into a single submission, as discussed in #191 

Most changes are independent and localized, but they collectively improve the experience of working with modern GLSL shader codebases.

### CODE REVIEW

File: src/main/java/glslplugin/formatter/GLSLFormattingBlock.java [hunk 1]
```diff
@@ -53,7 +63,51 @@ protected GLSLFormattingBlock(@NotNull ASTNode astNode, @Nullable Wrap wrap, @Nu
...
+        ASTNode child = node.getFirstChildNode();
+        while (child != null && childIndex < newChildIndex) {
+            IElementType childType = child.getElementType();
+            if (childType == GLSLTokenTypes.LEFT_PAREN) {
+                afterLeftParen = true;
+            } else if (childType == GLSLTokenTypes.RIGHT_PAREN) {
+                return false;
+            }
+
+            if (child.getText().trim().length() > 0) {
```
Comment (Darkyenus) [inline src/main/java/glslplugin/formatter/GLSLFormattingBlock.java:95]: This could maybe be a bit better, i.e. to avoid garbage and to not process the left paren. But it looks like this is not a hot function, so it is just an observation.

File: src/main/java/glslplugin/lang/elements/declarations/GLSLVariableDeclaration.java [hunk 2]
```diff
@@ -49,12 +49,11 @@ public GLSLVariableDeclaration(@NotNull ASTNode astNode) {
...
     @Override
     public boolean processDeclarations(@NotNull PsiScopeProcessor processor, @NotNull ResolveState state, @Nullable PsiElement lastParent, @NotNull PsiElement place) {
-        for (GLSLDeclarator declarator : getDeclarators()) {
-            if (PsiTreeUtil.isAncestor(lastParent, declarator, false))
-                break;// Can't see later-defined stuff
-
-            if (!declarator.processDeclarations(processor, state, null, place))
+        final GLSLDeclaratorList list = findChildByClass(GLSLDeclaratorList.class);
+        if (list != null) {
+            if (!list.processDeclarators(processor, state, lastParent, place)) {
```
Comment (Darkyenus) [inline src/main/java/glslplugin/lang/elements/declarations/GLSLVariableDeclaration.java:54]: This is nice.

**Review by Darkyenus — APPROVED** [review]
Thank you!
